Feature: Improve the lesson preview

cleicar opened the pull request:
<!-- Thank you for taking the time to contribute to The Odin Project. In order to get a pull request (PR) closed in a reasonable amount of time, you must include a baseline of information about the changes you are proposing. Please read this template in its entirety before filling it out to ensure that it is filled out correctly. -->

Complete the following REQUIRED checkboxes:
<!-- While editing this template, replace the whitespace between the square brackets with an 'x', e.g. [x] -->
-   [x] I have thoroughly read and understand [The Odin Project Contributing Guide](https://github.com/TheOdinProject/theodinproject/blob/main/CONTRIBUTING.md)
-   [x] The title of this PR follows the `keyword: brief description of change` format, using one of the following keywords:
  - `Feature` - adds new or amends existing user-facing behaviour
  - `Chore` - changes that have no user-facing value, refactors, dependency bumps, etc
  - `Fix` - bug fixes

Complete the following checkbox ONLY IF it is applicable to your PR. You can complete it later if it is not currently applicable:
-   [x] I have verified all tests and linters pass against my changes, and/or I have included automated tests where applicable

<hr>

**1. Because:**
<!--
If this PR closes an open issue, replace the XXXXX below with the issue number, e.g. Closes #2013. Or if the issue is in another TOP repo replace the #XXXXX with the URL of the issue, e.g. Closes https://github.com/TheOdinProject/curriculum/issues/XXXXX

Otherwise, provide a clear and concise reason for your pull request, e.g. what problem it solves or what benefit it provides. If this PR is related to, but does not close, another issue or PR, you can also link it as above without the 'Closes' keyword, e.g. "Related to #2013".
 -->
Closes https://github.com/TheOdinProject/theodinproject/issues/3073


**2. This PR:**
<!--
A bullet point list of one or more items outlining what was done in this PR to solve the problem(s) or implement the feature/enhancement.
 --> Creates a table to save the lesson previews
 --> When the user clicks the share button, it saves the content to the database associated to a `uuid` and copies the sharable url to clipboard.
--> When the user opens the link in another page, it retrieves the content from database and show in the page.

cleicar commented on db/schema.rb:
```diff
@@ -10,10 +10,11 @@
 
   # These are extensions that must be enabled in order to support this database
   enable_extension "plpgsql"
+  enable_extension "pgcrypto"
```
This is necessary to enable using `uuid` as the database id.

cleicar commented:
@KevinMulhern this is ready for review

KevinMulhern commented on app/javascript/components/lesson-preview/index.jsx:
```diff
@@ -8,9 +9,7 @@ import LessonContentInput from './components/lesson-content-input';
-
-const LessonPreview = () => {
+const LessonPreview = ({previewContent}) => {
   const [content, setContent] = useState('');
```
We can remove the useEffect for setting the content if the query is present and just set the previewContent as the initial content. It should be an empty string if it's empty.
```suggestion
  const [content, setContent] = useState(previewContent);
```

KevinMulhern commented on app/controllers/lessons/previews_controller.rb:
```diff
@@ -1,8 +1,20 @@
+    end
 
     def create
+      preview_link = LessonPreview.create(lesson_preview_params)
```
I don't think the error branch will ever be called with this since create will always return the LessonPreview object. We'll need to do something like this instead:

```ruby
    def create
      preview_link = LessonPreview.new(lesson_preview_params)
        
      if preview_link.save
        render json: { preview_link: preview_link(preview_link.id) }, status: :created
      else
        render json: { errors: preview_link.errors.full_messages }, status: :unprocessable_entity
      end
    end
```

KevinMulhern commented on app/controllers/lessons/previews_controller.rb:
```diff
@@ -1,8 +1,20 @@
+      preview_link = LessonPreview.create(lesson_preview_params)
+
+      if preview_link
+        render json: { preview_link: preview_link(preview_link.id) }
```
Can we inline the link please?
```suggestion
        render json: { preview_link: lessons_preview_url(uuid: preview_link.id)  }
```

KevinMulhern reviewed (changes requested):
This is looking really good 🔥 

I'm still a little undecided with the model name. `LessonPreview` does make sense, but I think `SharedPreview` might reveal the intention more. I'll have more of a think on it and get back to you.

I have a few more minor suggestions in mind after the `ShareButton` component is in place. But nothing major.

Thank you for all the work on this 💪

JuanVqz commented on db/migrate/20221021183414_create_lesson_previews.rb:
```diff
@@ -0,0 +1,8 @@
+class CreateLessonPreviews < ActiveRecord::Migration[6.1]
+  def change
+    create_table :lesson_previews, id: :uuid do |t|
+      t.text :content
```
@cleicar, I added this constraint, so, we need to add a validation model level

JuanVqz commented on app/javascript/components/lesson-preview/index.jsx:
```diff
@@ -29,15 +28,17 @@ const LessonPreview = () => {
 
-  const handleClick = () => {
-    navigator.clipboard.writeText(link).then(() => setCopied(true));
+  const handleClick = async () => {
```
@cleicar, we started moving the button to a ShareButton component, you can continue checking it, please

cleicar commented:
@KevinMulhern thanks for you review. We've addressed your comments. 

My two cents on the model name: `SharedPreview ` to me means more like a lesson that already  exists and is shared across courses. 

Either way, I'm fine with changing the name to whatever you decide. Thanks!

KevinMulhern commented on app/javascript/components/lesson-preview/components/share-button.jsx:
```diff
@@ -0,0 +1,30 @@
+
+  return (
+    <button type="button" className={`button ${classes}`} onClick={handleOnClick} >
+      {text}
```
Nit: can we use the copied conditional here instead of passing the text in as a prop please?
```suggestion
      {copied ? 'Copied!' : 'Share'}
```

All the copy/share logic can be within this component instead of the parent. The only thing I think we'll need to pass in as a prop is the content.

KevinMulhern commented on spec/models/lesson_preview_spec.rb:
```diff
@@ -0,0 +1,5 @@
+require 'rails_helper'
+
+RSpec.describe LessonPreview, type: :model do
+  it { is_expected.to validate_presence_of(:content) }
```
Lets add a test for the max length validation too:
```suggestion
  it { is_expected.to validate_presence_of(:content) }
  it { is_expected.to validate_length_of(:content).is_at_most(70_000) }
```

KevinMulhern commented on app/javascript/components/lesson-preview/index.jsx:
```diff
@@ -67,16 +62,19 @@ const LessonPreview = () => {
-        >
-          {copied ? 'Copied!' : 'Share'}
-        </button>
+        <ShareButton text={copied ? 'Copied!' : 'Share'} classes={copied ? 'button--secondary' : 'button--primary'} onClick={handleShareClick}>
```
Can we only show the button if there is any content please?

```jsx
const hasContent = content.length > 0;
...
 { hasContent && <ShareButton .... /> }

JuanVqz commented on app/javascript/components/lesson-preview/index.jsx:
```diff
@@ -1,26 +1,24 @@
-
-const LessonPreview = () => {
+const LessonPreview = ({previewContent}) => {
   const [content, setContent] = useState('');
```
if I set the `content` with `previewContent` by default it returns the following error:

```
Warning: `value` prop on `textarea` should not be null. Consider using an empty string to clear the component or `undefined` for uncontrolled components.
```
Am I doing something wrong?

KevinMulhern replied:
It will be the `@preview.content` on this line causing that. A new record will have nil content.
```ruby
<%= react_component('lesson-preview/index', { previewContent: @preview.content }) %>`
```

But we can call `to_s` on it to ensure we always get a string, even when content is nil.
```ruby
<%= react_component('lesson-preview/index', { previewContent: @preview.content.to_s }) %>
```

JuanVqz replied:
Yep I noticed that I fixed it already 

Thank you!

KevinMulhern commented on db/migrate/20221021183414_create_lesson_previews.rb:
```diff
@@ -0,0 +1,8 @@
+class CreateLessonPreviews < ActiveRecord::Migration[6.1]
+  def change
+    create_table :lesson_previews, id: :uuid do |t|
+      t.text :content, default: '', null: false
```
Can we remove the default here please? we don't ever want to have lesson previews with empty content.

KevinMulhern commented on app/javascript/components/lesson-preview/index.jsx:
```diff
@@ -1,26 +1,22 @@
-
-const LessonPreview = () => {
-  const [content, setContent] = useState('');
+const LessonPreview = ({ previewContent }) => {
```
Nit: Can we rename `previewContent` to `sharedContent` please? 

It's difficult to discern the difference between `previewContent` and `content` within these components by their names which could lead to some confusion. I think `sharedContent` will make it clearer what the passed in content represents.

KevinMulhern commented on app/javascript/components/lesson-preview/components/share-button.jsx:
```diff
@@ -0,0 +1,39 @@
+const ShareButton = ({ content }) => {
+  const [copied, setCopied] = useState(false);
+
+  useEffect(() => {
```
Suggestion: Instead of switching back to "Share" after 4 seconds, lets only switch when/if the content changes again:

```jsx
useEffect(() => {
  if (copied) {
    setCopied(false);
  }
}, [content]);
```

KevinMulhern commented on app/models/lesson_preview.rb:
```diff
@@ -0,0 +1,3 @@
+class LessonPreview < ApplicationRecord
+  validates :content, presence: true, length: { maximum: 70_000 }
```
Nit: To validate this length on the client side, can we add a `maxLength={10}` attribute to the text field [here](https://github.com/TheOdinProject/theodinproject/blob/main/app/javascript/components/lesson-preview/components/lesson-content-input.jsx#L10) please?

KevinMulhern reviewed (changes requested):
Thanks you all the work you've both put into this.

This is almost there, I think we just need a couple more things to harden this up for production:

We need to rate limit the share button to ensure it cannot be abused and to avoid us ending up with a lot of unnecessary lesson previews. We can do that with [rack attack here](https://github.com/TheOdinProject/theodinproject/blob/main/config/initializers/rack_attack.rb#L4). 

Lesson previews will be temporary, we want to remove any that are a month or more old. We'll need an expired scope on `LessonPreview` and a new [job](https://github.com/TheOdinProject/theodinproject/tree/main/app/jobs) to do that.

JuanVqz commented:
Hi @KevinMulhern! did I miss something here? happy to solve it.

KevinMulhern commented:
Hey @JuanVqz, sorry for the confusion.

Just these couple of things are remaining before we can merge:
https://github.com/TheOdinProject/theodinproject/pull/3454#pullrequestreview-1160989208

> This is almost there, I think we just need a couple more things to harden this up for production:
> 
> We need to rate limit the share button to ensure it cannot be abused and to avoid us ending up with a lot of unnecessary lesson previews. We can do that with [rack attack here](https://github.com/TheOdinProject/theodinproject/blob/main/config/initializers/rack_attack.rb#L4).
> 
> Lesson previews will be temporary, we want to remove any that are a month or more old. We'll need an expired scope on `LessonPreview` and a new [job](https://github.com/TheOdinProject/theodinproject/tree/main/app/jobs) to do that.

JuanVqz commented on app/services/lesson_previews_remover.rb:
```diff
@@ -0,0 +1,13 @@
+class LessonPreviewsRemover
+
+  def call
+    more_than_one_month_lesson_previews_old.destroy_all
```
do you think we should set the batch_size manually? I guess the default is 1000

KevinMulhern commented on app/services/lesson_previews_remover.rb:
```diff
@@ -0,0 +1,11 @@
+class LessonPreviewsRemover
```
Appologies @JuanVqz, I think we got our wires crossed with whats needed here. We need a job instead of a service. Something very similar to what we have for discarding projects https://github.com/TheOdinProject/theodinproject/blob/main/app/jobs/discard_project_submission_job.rb 

```ruby
class RemoveExpiredPreviewsJob
  include Sidekiq::Worker
  sidekiq_options retry: 1, dead: false

  def perform
    LessonPreview.expired.each(&:destroy)
  end
end
```

Then we'll need a rake task to trigger the job everyday using Cron on Heroku: https://github.com/TheOdinProject/theodinproject/blob/main/lib/tasks/project_submissions.rake

JuanVqz replied:
I wanted to use that service in the job. but okay, I'll put it all together in the job's body, :+1:

JuanVqz commented:
hey, @KevinMulhern this is ready to be tested again, could you please do it? Thanks for the patience :+1:

KevinMulhern reviewed (commented):
QA'd and all works great, thanks for all the work you've put into this and your patience throughout this review.

All that remains is just one small change to fix one of my earlier suggestions, sorry for the trouble @JuanVqz

JuanVqz commented:
> QA'd and all works great, thanks for all the work you've put into this and your patience throughout this review.
> 
> 
> 
> All that remains is just one small change to fix one of my earlier suggestions, sorry for the trouble @JuanVqz 

No problem, I liked being reviewed on the fly... I'll fix it tomorrow or you can merge your suggestion I don't have enough permission to do it.
I'd like to contribute more to this project, thanks! 💎

JuanVqz commented:
Hey, @KevinMulhern I did fix the requested changes, could you please review it again?

btw, don't forget to add the cron job in Heroku `lesson_previews:destroy_expired`

Thanks!

JuanVqz commented:
I have two scenarios that we may improve:

1. what about if a user is working on something but spent over a month (but still working on it), then we remove it since the task was taking into account the created_at date, I think it will be better if we take into account the updated_at attribute.

2. what about if a user has a link Eg. `http://localhost:3000/lessons/preview?uuid=5e49d961-9bda-4819-87da-73e5f4f8628a` and bookmark it or saves it to review later, but a month already passed, the user will visit the page and it will be blank because the record was removed, should we add an alert saying something?

Thanks

KevinMulhern commented:
Sorry for the delay @JuanVqz, I haven't had much time available TOP this past week.

> 1. what about if a user is working on something but spent over a month (but still working on it), then we remove it since the task was taking into account the created_at date, I think it will be better if we take into account the updated_at attribute.

This would be getting a little out of scope of what the preview tool is designed for at the moment and into document management territory imo. Ideally contributors will be storing their work in a md file in our curriculum repo and only using the preview tool to make sure the content they are producing works as intended with our styling. I wouldn't entirely exclude this idea though, there may be a use case for it in the future as we see how this new and improved share functionality is used by contributors.

> 2. what about if a user has a link Eg. http://localhost:3000/lessons/preview?uuid=5e49d961-9bda-4819-87da-73e5f4f8628a and bookmark it or saves it to review it later, but a month already passed, the user will visit the page and it will be blank because the record was removed, should we add an alert saying something?

Good call, I think it would be worth adding a flash to communicate the preview doesn't exist or may have expired.

KevinMulhern reviewed (approved):
Thank you for all the work you've put into this @cleicar @JuanVqz. This will be a much more useful tool because of it 💪 🏆
